### PR TITLE
Test plan data cleanup

### PR DESCRIPTION
**Pull request by tharindu1992**
**Purpose**

Resolves:#1040 

**Goals**

Clean up data of test plans including data base entries in test_plan, test_case, test_scenario and log files in S3

**Approach**

Added new functionality to cleanup sub command and through that user can delete all test plan related data keeping only a specific amount of builds for each infra combination in each job. Graph data is not deleted because all previous emails that contain the graphs are still been used.  

**Security checks**
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java [hunk 1]
```diff
@@ -50,13 +65,54 @@ public static String getS3BucketURL() {
...
 
+    public static void deleteTestPlan(TestPlan testPlan) {
+
+        try {
+            ArtifactReadable artifactDownloadable = new AWSArtifactReader(ConfigurationContext.
+                    getProperty(ConfigurationContext.ConfigurationProperties.AWS_REGION_NAME),
+                    ConfigurationContext.getProperty(ConfigurationContext.ConfigurationProperties.AWS_S3_BUCKET_NAME));
+
+            String s3KeyName = deriveS3TestPlanDirPath(testPlan, artifactDownloadable);
+            logger.info("Files cleaned in :" + s3KeyName);
```
Comment (kasunbg) [inline common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java]:
```suggestion
            logger.info("Started to clean artifacts of test-plan: " + testPlan.getId() + ", s3 bucket: " + s3KeyName);
```

File: common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java [hunk 2]
```diff
@@ -50,13 +65,54 @@ public static String getS3BucketURL() {
...
+            String clientRegion = ConfigurationContext.getProperty
+                    (ConfigurationContext.ConfigurationProperties.AWS_REGION_NAME);
+            String bucketName = ConfigurationContext.getProperty
+                    (ConfigurationContext.ConfigurationProperties.AWS_S3_BUCKET_NAME);
+
+            AmazonS3 s3Client = AmazonS3ClientBuilder.standard()
+                    .withCredentials(new ProfileCredentialsProvider())
```
Comment (kasunbg) [inline common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java]: We have been using the AWS SDK's PropertiesFileCredentialsProvider for credentials. Why this change?

https://github.com/wso2/testgrid/blob/46f29344e230807f556eefd356fc8f5c030e4267/infrastructure/src/main/java/org/wso2/testgrid/infrastructure/providers/aws/AMIMapper.java#L71

File: common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java [hunk 3]
```diff
@@ -50,13 +65,54 @@ public static String getS3BucketURL() {
...
+                    .withRegion(clientRegion)
+                    .build();
+
+            ObjectListing objectListing = s3Client.listObjects(bucketName, s3KeyName);
+            List<S3ObjectSummary> objectSummaries = objectListing.getObjectSummaries();
+
+            for (S3ObjectSummary os : objectSummaries) {
+                logger.info(StringUtil.concatStrings("deleting file", os.getKey(), "of test plan ",
```
Comment (kasunbg) [inline common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java]:
```suggestion
                logger.debug(StringUtil.concatStrings("deleting file ", os.getKey(), " of test plan ",
```

File: common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java [hunk 4]
```diff
@@ -50,13 +65,54 @@ public static String getS3BucketURL() {
         return String.join("/", TestGridConstants.AMAZON_S3_URL, s3BucketName);
     }
 
+    public static void deleteTestPlan(TestPlan testPlan) {
```
Comment (kasunbg) [inline common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java]:
```suggestion
    public static boolean deleteTestPlan(TestPlan testPlan) {
```

I think you need to return the deletion success/failure status.

File: common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java [hunk 5]
```diff
@@ -50,13 +65,54 @@ public static String getS3BucketURL() {
...
+            }
+
+            s3Client.deleteObject(new DeleteObjectRequest(bucketName, s3KeyName));
+
+        } catch (AmazonServiceException e) {
+            logger.error("Amazon S3 couldn't process request ", e);
+        } catch (SdkClientException e) {
+            logger.error("Amazon S3 couldn't be contacted for a response" , e);
+        } catch (ArtifactReaderException e) {
+            logger.error("Error occurred when reading the artifact.", e);
```
Comment (kasunbg) [inline common/src/main/java/org/wso2/testgrid/common/util/S3StorageUtil.java]: Missing context. You need to log for which test-plan this issue occurred, and any other useful information.

Suppose, you have a 1000MB log file where this issue was there in the middle. Think how much context information you need to tackle down and fix the issue or identify the affects of it.

File: core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java [hunk 6]
```diff
@@ -66,11 +71,16 @@
...
     private String grafanaApikey =
             ConfigurationContext.getProperty(ConfigurationContext.ConfigurationProperties.GRAFANA_APIKEY);
-
     private List<String> toDelete = new ArrayList<String>();
 
+    @Option(name = "--count",
+            usage = "Build Count",
+            aliases = { "-c" },
+            required = true)
+    private int remainingBuildCount = 100;
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java]: I think we need to re-look at this API from the user's perspective. I do not think an end user would understand what is meant by 'count'. The usage of this API is as follows:

`testgrid cleanup --count=10`

This feels like we are going to 'clean' 10 items. But you actually meant to say, how many builds to keep.

File: core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java [hunk 7]
```diff
@@ -116,13 +126,21 @@ public void execute() throws CommandExecutionException {
             List<String> allTestPlans = testPlanUOW.deleteDatasourcesByAge(remainingBuildCount);
             logger.info("number of items to delete: " + allTestPlans.size());
 
+            logger.info("Clearing S3 files");
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java]: Missing context info. Shall we print how many items user wants to keep?

File: core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java [hunk 8]
```diff
@@ -116,13 +126,21 @@ public void execute() throws CommandExecutionException {
             List<String> allTestPlans = testPlanUOW.deleteDatasourcesByAge(remainingBuildCount);
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java]: This API is not correct. Why does an API like TestPlanUOW#deleteDatasourcesByAge return a list of test plans? Isn't it supposed to delete and return a true/false status?

Have you thought about memory requirements to load all old test plans?

File: core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java [hunk 9]
```diff
@@ -116,13 +126,21 @@ public void execute() throws CommandExecutionException {
...
             logger.info("number of items to delete: " + allTestPlans.size());
 
+            logger.info("Clearing S3 files");
+            for (String deletingTestPlan : allTestPlans) {
+                deleteS3(deletingTestPlan);
+            }
+
+            logger.info("Deleting test plan from DB");
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java]: Which test-plan? Again, missing context. Btw, I think think should be a debug log.

File: core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java [hunk 10]
```diff
@@ -131,6 +149,8 @@ public void execute() throws CommandExecutionException {
             }
 
         } catch (IllegalArgumentException e) {
+            throw new CommandExecutionException(e.getMessage(), e);
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java]: You are catching an IllegalArgumentException. Do you know which argument it was? Shouldn't we log it? Exception handling is really important for long-term maintenance of a project.

File: core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java [hunk 11]
```diff
@@ -232,6 +252,19 @@ private void clearDataSources(String datasource) {
...
 
+    /**
+     * This method delete the files in s3 for a given test plan
+     * @param testPlanId tes plan id of which the files need to be deleted
+     * @throws TestGridDAOException
+     */
+    public void deleteS3(String testPlanId) throws TestGridDAOException {
+
+        Optional<TestPlan> testPlanEntity = testPlanUOW.getTestPlanById(testPlanId);
+        TestPlan testPlan = testPlanEntity.get();
```
Comment (kasunbg) [inline core/src/main/java/org/wso2/testgrid/core/command/CleanUpCommand.java]: Possible cause for NPE. In case the test-plan got deleted from the db by some other means, this call will fail. We need to pay attention concurrency situations.

File: core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java [hunk 12]
```diff
@@ -23,19 +29,54 @@
 public class CleanUpCommandTest extends PowerMockTestCase {
```
Comment (kasunbg) [inline core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java]: Good we are writing tests. What's the code coverage of this one?

File: core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java [hunk 13]
```diff
@@ -23,19 +29,54 @@
...
+        deploymentPattern.setName("default-" + randomStr);
+        deploymentPattern.setProduct(product);
+
+
+        testPlan.setId("TP1");
+        testPlan.setTestRunNumber(1);
+        testPlan.setDeploymentPattern(deploymentPattern);
+        testPlan.setInfraParameters(infraParamsString);
+        testPlan.setDeployerType(TestPlan.DeployerType.SHELL);
+        testPlan.setScenarioTestsRepository(Paths.get(workspaceDir, "/workspace/scenarioTests").toString());
```
Comment (kasunbg) [inline core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java]: Incorrect use of Paths API.

```suggestion
        testPlan.setScenarioTestsRepository(Paths.get(workspaceDir, "workspace", "scenarioTests").toString());
```

File: core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java [hunk 14]
```diff
@@ -23,19 +29,54 @@
...
+        testPlan.setInfrastructureRepository(Paths.get(workspaceDir, "/workspace/infrastructure").toString());
+        testPlan.setDeploymentRepository(Paths.get(workspaceDir, "/workspace/deployment").toString());
+        testPlan.setKeyFileLocation(Paths.get(workspaceDir, "/workspace/testkey.pem").toString());
```
Comment (kasunbg) [inline core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java]: Same as above.

File: dao/src/main/java/org/wso2/testgrid/dao/repository/TestPlanRepository.java [hunk 15]
```diff
@@ -392,6 +392,22 @@ public TestPlan getLastFailure(TestPlan testPlan) {
...
+    public void deleteTestPlans(List<String> testPlans) {
+        String sql;
+        EntityTransaction txn = entityManager.getTransaction();
+        txn.begin();
+
+        for (String testplan : testPlans) {
+            sql = "delete from test_plan where id = '" + testplan + "'";
+            entityManager.createNativeQuery(sql).executeUpdate();
+        }
+        txn.commit();
```
Comment (kasunbg) [inline dao/src/main/java/org/wso2/testgrid/dao/repository/TestPlanRepository.java]: Shouldn't this go under a try-catch block? 

It is important because if the line #406 threw an exception, you need to rollback the transaction. ex: https://www.boraji.com/hibernate-5-and-jpa-2-persist-find-merge-and-remove-example

File: core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java [hunk 16]
```diff
@@ -23,19 +29,54 @@
...
+
+
+        this.product = new Product();
+        product.setName(productName);
+
+        this.deploymentPattern = new DeploymentPattern();
+        deploymentPattern.setName("default-" + randomStr);
+        deploymentPattern.setProduct(product);
+
+
```
Comment (kasunbg) [inline core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java]:
```suggestion
```

Unnecessary new lines.

File: core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java (hunk 12 above)
Comment (tharindu1992) [reply core/src/test/java/org/wso2/testgrid/core/command/CleanUpCommandTest.java]: Method: 80%
Line 54%
